src/sites/twitch-twilight/modules/chat/input.jsx

```diff
@@ -647,21 +647,49 @@ export default class Input extends Module {
 
 
 	// eslint-disable-next-line class-methods-use-this
-	sortFavorites(results) {
-		if (!this.chat.context.get('chat.tab-complete.prioritize-favorites')) {
-			return results;
+	sortEmotes(emotes, input) {
+		const preferFavorites = this.chat.context.get('chat.tab-complete.prioritize-favorites');
+		const lowerCaseInput = input.toLowerCase();
+
+		const locale = new Intl.Collator();
+		const localeCI = new Intl.Collator(undefined, {sensitivity: 'accent'}); // case insensitive
+
+		const startsWithInput = new Set();
+		const startsWithInputCI = new Set(); // case insensitive
+		for (let i = 0; i < emotes.length; ++i) {
+			const str = emotes[i].replacement;
+			if (str.startsWith(input))
+				startsWithInput.add(str);
+			else if (str.toLowerCase().startsWith(lowerCaseInput))
+				startsWithInputCI.add(str);
 		}
 
-		return results.sort((a, b) => {
-			if (a.favorite) {
-				return b.favorite ? a.replacement.localeCompare(b.replacement) : -1;
-			}
-			else if (b.favorite) {
-				return 1;
-			}
-			else {
-				a.replacement.localeCompare(b.replacement)
-			}
+		return emotes.sort((a, b) => {
```

Comment:
One last note. I'd like to introduce a new setting to select between the new sorting behavior vs the old sorting behavior. Users may be used to how things work already and not particularly want to adapt.

Reply:
Do you think the new behavior should be opt-in, or the default?

Reply:
Also, looking into the changes you suggested made me realize that I didn't properly consider how emoji should be sorted against twitch emotes in my original commit. I think that for people using the new behavior it should work like this:

If the user has `'chat.tab-complete.emotes-without-colon'` enabled, then emoji should always appear at the top of the list. Currently, emoji are only returned when the user input starts with `:`, so for users who have tab-completion enabled, I think it's a safe bet that if they start their input with `:`, it's because they want an emoji.

If the user does not have tab-completion without colon enabled, then emoji should appear at the bottom of the list, as is the current behavior. This logic here is that users who are used to typing `:ka` for `Kappa` shouldn't need to scroll through `:kazakhstan:`, `:kaaba:`, and `:kangaroo:` every time.

Then again, an argument could be made that emoji should always be put at the top of the list, since someone who doesn't want emoji can simply disable tab completion for emoji in the input settings. I lean against this point of view, but I thought I'd bring it up. Let me know what you think.

Reply:
In *my* opinion it should be opt-in because it's quite a big change for users' muscle memory on how to tab-complete emotes. (Also to not mess with the native way it works if they switch back to mobile and don't have FFZ installed there)

A bit of background as to why the current tab-completion is the way it is:

Twitch has it's own native tab-completion that FFZ hooks into and adds extension emotes into.
The priority of it is:
- Twitch Global Emotes
- Twitch Sub Emotes
- FFZ Emotes
- Emotes added by other addons / extensions that utilize the FFZ emote system

On-top of that, Twitch supports tab-completing sub-emotes without the prefix.
So say you have a lot of "Sip" or "Love" emotes, you can just type `:love` and it will show all the sub-emotes that start with `love`, next to global emotes that start with `love`.

Now... this is the same system that FFZ uses, since Twitch's comparison is being used.
This is why `KappaClaus` is preferred before everything else ("prefix" `Kappa` removed, and sorted to the top because it's a global emote) and why `peepoClap` is in 2nd place (`peepo` being the "prefix" here)

Reply:
> One last note. I'd like to introduce a new setting to select between the new sorting behavior vs the old sorting behavior. Users may be used to how things work already and not particularly want to adapt.

Addressed in 0ea064a.